README.md

# UAF Auction program

This program are to be used during an auction or a fleamarket by an aquarium club or similar organisation.

1. It can handle registration of sellers and the registration of items to sell. This can be categorised by categories defined in a database.

2. It can generate labeles (pdf) and economic reports.

3. It registers price of items sold as well as if it was sold at auction or fleamarket.



## Requirements

This is a client server program written in Python 3.5, HTML5, JavaScript and CSS with Bootstrap. 
It is meant to be run on a Linux (tested on Ubuntu) server with Apache, but should run on other python compatible servers.

### Install and setup of webserver on Linux (Ubuntu / Raspbian)
Tested on Ubuntu 14.04, 16.04, Linux Mint 18, Raspbian 2017-03-02-raspbian-jessie-lite and 
piOS Raspbian 11 Bullseye (2022-11-22).
.

This file describes the installation on a Raspberry pi with Raspbian 11 in the folder /home/pi/uaf.

Install apache with Python3 support:

```
sudo apt-get update
sudo apt-get upgrade

sudo apt-get install apache2
sudo apt-get install libapache2-mod-wsgi-py3 python3-dev
```

Enable mod_wsgi, need for running python under apache.
```
sudo a2enmod wsgi 
```

Download the code from gitlab, requires ssh-keys to access gitlab:
```
sudo apt-get install git
git clone git@gitlab.com:kristianpersson/uaf_auction.git
mv uaf_auction/ uaf/
```
Or copy the code in another way, but place it in /home/pi/uaf.


For apache 2.4 the uaf.conf file should look like this, if the program is in the folder /home/pi/uaf.
```
<VirtualHost *:80>

                WSGIDaemonProcess uaf user=pi group=pi threads=5 home=/home/pi/uaf
                WSGIScriptAlias / /home/pi/uaf/uaf.wsgi

                <Directory /home/pi/uaf/>
                     WSGIProcessGroup uaf
                     WSGIApplicationGroup %{GLOBAL}

                     WSGIScriptReloading On
                     Require all granted
                </Directory>

                LogLevel warn
                CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
```

If the server has an domain name (like auction.uaf.se) and the program is to be accessed via that, then 
add the line:

```
ServerName auction.uaf.se
```
As the first line after <VirtualHost *:80>.
 

Copy the conf-file to /etc/apache2/sites-available\
```
sudo cp uaf.conf /etc/apache2/sites-available/
```

Deactivate the default website, and activate the flask-site.
```
sudo a2dissite 000-default.conf 
sudo a2ensite uaf.conf
sudo systemctl reload apache2
```

There must be a wsgi file in the same folder as the python program. 
In the above VirtualHost definifion WSGIScriptAlias points to this file. 
This links the requested path in the webb-browser to the python program to run. 
It should look like this:
```
import sys

sys.path.append('/home/pi/uaf')
from uaf import app as application
```


### Python3

If needed:
```
sudo apt-get install python3-pip 
```

Always upgrade pip to the latest version:
```
sudo pip3 install --upgrade pip
```

Install build-essentials 
```
sudo apt-get install build-essential
```

Install rust for compilation of bcrypt. Rust must be version 1.56 or later.
On raspberry pi I did not manage to get rust working, so use an older version of bcrypt 3.2.2 
which is not dependent on rust.

```
curl https://sh.rustup.rs -sSf | sh
```

Needed Python libraries:

* flask
* flask_login
* reportlab
* bcrypt

Install with pip:
```
sudo pip3 install flask_login
sudo pip3 install reportlab
sudo pip3 install bcrypt
or
sudo pip3 install bcrypt==3.2.2
```
In order to install bcrypt on Ubuntu 14.04 and Rasbian, libffi-dev was needed:
```
sudo apt-get install  libffi-dev
```

If internal server error, check the error log for error messages:
```
tail -n 20 /var/log/apache2/error.log
```

If PIL error in logfile:
```
>>> from PIL import Image
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/PIL/Image.py", line 100, in <module>
    from . import _imaging as core
ImportError: libopenjp2.so.7: cannot open shared object file: No such file or directory
```
Uninstall and reinstall of Pillow might solve the problem:
```
sudo pip3 uninstall Pillow
sudo pip3 install Pillow
```

### Configure the program
#### Config file
The program has a config file that gives the location and name of the database.
Copy the example config file and edit the content.
```
cp config.cfg.example config.cfg
```

Content of config file:
```
SECRET_KEY='a secret string'
DATABASE='auktion_generated.db3'
GMAILUSER='username@gmail.com'
GMAILPASSWORD='password'
```

The most important is DATABASE which is the name of the database.
SECRET_KEY generates the key for cookies and should be a random string.
GMAILUSER and GMAILPASSWORD was used to send notification emails from this gmail account when users are 
created and posts are added. In Settings under Administration menu on the webinterface the reciving mailaddress
can be given, empty means no emails are sent.
The mail function does not work anymore as Google has increased the security for Google accounts.

#### Database
The program needs a database to work. A new example database can be generated with:
```
python3 make_database.py
```
The admin user in that database is Kalle Persson. Username kalle.persson@mail.com and password password.
The database can be reset by creating a new event. All existing users will then be removed,
including the admin and a new admin will be created.


CHANGELOG
Version 0.85 2022-11-22
    Updated the README.md to get a working install on Raspbian 11 (Bullseye).
    Added a uaf.conf for apache2.

Version 0.84 2019-11-02
    Fixed bug when admin registered posts for another user. Causing 500 error.
    Fixed typo in welcome screen.

Version 0.83 2019-10-09
    Wrong path to images for pdf generation, causing server error 500.

Version 0.82 2019-10-02
    Fixed bug when generating individual sales reports.
    Fixed bugs when generating database.
    Fixed bug when createing event.

Version 0.81 2018-11-21
    Added the possibility to "close" a seller that want to leave early. The seller and all posts are marked as closed
    and a warning is shown in auction and fleamarket pages when trying to buy objects from the seller. Admin can print
    economic compilation reports for indiviual sellers in the handle sellers page. This needed the addition of the field
    "is_closed" to tables posts and sellers.

Version 0.80 2018-11-18
    Added a new label format 65 * 34 mm.
    Added border / no border options in the label setup and a corresponding border collumn in the label_type table.
    Added option to list sellers posts on "show info about seller" page.
    Made a new logo image to the labels, better resolution 250 * 250 pixels instead of 64 * 64.

Version 0.79 2018-10-25
    Removed quantity from some other places as edit posts, list my posts, register posts for other seller.

Version 0.78 2018-10-25
    Removed quantity from register posts.
    Minor tweaks to labels.

Version 0.77 2018-10-23
    Issue 63. Changed seller to buyer in text at the fleamarket page.
    Issue 64. Added total mount to pay at the bottom of the fleamarket page.

Version 0.76 2018-10-23
    New label design.
    Support for three different 3 by 8 lable types. Admin can select label type on a new label type admin page.

Version 0.75 2018-03-19
    Added quantity and price columns to "list my own posts" page.
    Added information about checked in posts on seller receipt.

Version 0.74 2018-03-09
    Changeing the event divided the provision by 100 each time and also left the event as closed because the registration_open field was left empty.
    Changed the images search text from länk to Bildsök.

Version 0.73 2018-03-09
    Fixed a bug in edit event that removed the input for scientific names from register posts, the database was not updated correctly.

Version 0.72 2018-02-28
    Fixed some small bugs in notification email. Added a test button for email.
    Added a calculate change function to flea market page.
    Fixed so that price change are updated on the fly and not only when the focus leaves the price input.

Version 0.71 2018-02-06
    Added error handling to the send email function.
    Issue 55. Registration of new sellers is turned off if the event is not open. Unless you are logged in as an admin, they can register new sellers even if the event is closed.
    Issue 60. Added quantity and description to the label.

Version 0.70 2018-02-05
    Issue 61. Added e-mail notifications via gmail. A mail is sent when a user is registered and when posts are registered. 
    The notification email address is set via the adminstrator interface and the gmail information is set in the config.cfg file.

Version 0.69 2018-02-01
    Fixed  bug where creating a new event did not add columns in the database for check in functionality.

Version 0.68 2018-01-22
    Issue 59. Added an unsold button to the auction page. Marks the post as unsold in the database, which is usefull for follow up of the auction.
    Added sellers all posts to the compilation, not just the sold posts. Easier for the seller to se what was sold.
    Each post is marked as checked in or not on the compilation.
    Removed some debug print statements.

Version 0.67 2018-01-18
    Issue 57 & 58 Added columns for check in on event for seller and posts in the database.
    Issue 57 & 58 Added pages for set check in status for sellers and posts. 
    Issue 58 Auction and fleamarket pages rises a warning when trying to sell a post that has not been checked in.
    Issue 58 Newly created posts are marked as not checked in.
    Issue 57 Newly created selles are marked as not checked in.
    Issue 57 Compilation pdf with only checked in sellers is available on the report page.

Version 0.66 2017-12-18
    Split the sellers page into one information page with address and such and a manage seller page, where admins can print, remove etc.
    Added sorting to the sellers tables.
    Issue 44. Admin can delete individual posts for all sellers.
    Issue 45. Admin can change password for seller.
    Issue 40. Added information if the scientific name is mandatory to the database, more flexible than the previous hack solution.
    Issue 41. When selecting non living things in the register posts page scientific name input is hidden, actually governed by a column in the database.
    Issue 46. Seller id number is added to the sellers pages.
    Issue 47. User edited posts are marked as unprinted.
    Issue 51. Added seller id to dropdown in register posts for other seller page.
    Issue 53. Limited the length of the description to 50 characters on the register posts page, make it possible to print description on label.
    Issue 54. Added quantity to registring posts, and editing posts.
    Selection of a auction type on the register page deletes the fleamarket price and vice versa.

Version 0.65 2017-12-12
    Added timeplots on auction and fleamarket sales.

Version 0.64 2017-11-30
    Updated the text on the receipt and compilation pdfs, also added logo.
    Issue 42. Added new pdf for fleamarket seller that calculates the total amount of money that the clerk has handled. The clerk is saved in the database.

Version 0.63
    Updated the README.md file to python 3.5
    Fixed some layout issues on the display pages.

Version 0.62 2017-11-28

faker.py
# coding=utf-8
__author__ = 'kristian'

import random
import time

class Faker:

    def __init__(self):
        """
        Generates data to insert into the database
        """
        self.fish_names = [
            [u"Ancistrus sp.", u"skäggmunsmal"],
            [u"Mikrogeophagus ramirezi", u"Fjärilsciklid"],
            [u"Pterophyllum scalare", u"Skalar"],
            [u"Betta splendens", u"Kampfisk"],
            [u"Poecilia reticulata", u"Guppy"],
            [u"Xiphophorus maculatus", u"Platy"],
            [u"Paracheirodon axelrodi", u"Kardinaltetra"],
            [u"Carassius auratus auratus", u"Guldfisk"],
            [u"Crossocheilus oblongus", u"Siamesisk algätare, algätare"],
            [u"Paracheirodon innesi", u"Neontetra"],
            [u"Glyptoperichthys gibbiceps", u"Segelfenspleco"],
            [u"Apistogramma cacatuoides", u"Kakadua ciklid"],
            [u"Labidochromis caeruleus", u"Golden labidochromis"],
            [u"Pelvicachromis pulcher", u"Palettciklid"],
            [u"Chromobotia macracanthus", u"Praktbotia"],
            [u"Poecilia sphenops", u"Blackmolly"],
            [u"Trichogaster lalius", u"Dvärggurami"],
            [u"Otocinclus vittatus", u"Randig otocinclus"],
            [u"Metriaclima estherae", u"Rödzebra"],
            [u"Hasemania nana", u"Koppartetra"],
            [u"Balantiocheilus melanopterus", u"Silverhaj"],
            [u"Xiphophorus helleri", u"Svärdbärare"],
            [u"", u""]
        ]

        self.plant_names = [
            [u"", u"Aponogeton rigidifolius"],
            [u"Hornsärv", u"Ceratophyllum demersum"],
            [u"Cryptocoryne", u"Cryptocoryne wendtii 'brown'"],
            [u"Dvärganubias", u"Anubias barteri var. nana"],
            [u"Svärdsplanta", u"Echinodorus 'Ozelot'"],
            [u"", u"Hygrophila difformis"],
            [u"Ciklidgräs", u"Ophiopogon japonicus"],
            [u"", u"Rotala rotundifolia"],
            [u"", u""]
        ]

        self.shrimp_names = [
            [u"Körsbärsräka", u"Neocaridina davidi var. red"],
            [u"Gul Körsbärsräka", u"Neocaridina davidi var. yellow"],
            [u"Röd Riliräka", u"Neocaridina davidi var. Röd Rili"],
            [u"Rödsakura Räka", u"Neocaridina davidi var. Red sakura"],
            [u"Zebraräka", u"Caridina cf. babaulti ssp. Indian zebra"],
            [u"Crystal red räka", u"Caridina cf. cantonensis ”Crystal Red”"]
        ]

        # Most common surnamnes 2016
        self.surnames = [
            u"Andersson",
            u"Karlsson",
            u"Johansson",
            u"Karlsson",
            u"Nilsson",
            u"Eriksson",
            u"Larsson",
            u"Olsson",
            u"Persson",
            u"Svensson",
            u"Gustafsson"
        ]

        # Most common first names in order 2016
        self.first_names = [
            u"Anna",
            u"Lars",
            u"Eva",
            u"Maria",
            u"Mikael",
            u"Anders",
            u"Johan",
            u"Karin",
            u"Per",
            u"Erik",
            u"Karl",
            u"Peter",
            u"Jan",
            u"Thomas",
            u"Kristina",
            u"Lena",
            u"Sara",
            u"Kerstin",
            u"Emma",
            u"Ingrid"
        ]

        # Most common street names in Sweden
        self.street_names = [ 
            u"Ringvägen",
            u"Storgatan",
            u"Skolgatan",
            u"Järnvägsgatan",
            u"Skogsvägen",
            u"Villavägen"
        ]

        # Largest cities in Sweden
        self.cities = [
            [u"Stockholm", u"16"],
            [u"Göteborg", u"41"],
            [u"Malmö", u"21"],
            [u"Uppsala", u"75"],
            [u"Västerås", u"72"],
            [u"Örebro", u"70"],
            [u"Linköping", u"58"],
            [u"Helsingborg", u"25"],
            [u"Jönköping", u"55"],
            [u"Norrköping", u"60"],
            [u"Umeå", u"90"],
            [u"Gävle", u"80"],
            [u"Borås", u"50"],
            [u"Eskilstuna", u"63"],
            [u"Södertälje", u"15"],
            [u"Karlstad", u"65"],
            [u"Täby", u"18"],
            [u"Växjö", u"35"],
            [u"Halmstad", u"30"]
        ]
        # Most common passwords 2016
        self.passwords = [
            u"123456",
            u"password",
            u"12345",
            u"12345678",
            u"football",
            u"qwerty",
            u"1234567890",
            u"1234567",
            u"princess",
        ]

        self.aquarium_clubs = [
            u"Västra Aros Akvarieförening",
            u"Uppsala Akvarieförening",
            u"Spånga Akvarieförening",
            u"Haninge Akvarieförening",
            u"Bollmora Akvarieklubb",
            u"Carlskrona Akvarieförening",
            u"Linköpings akvarieförening",
            u"Malmö Akvarieförening",
            u"Stockholms akvarieförening",
            u"Örebro Akvarieklubb",
            u"Umeå Akvarieförening"
        ]

        self.comments_animals = [
            u"Unga honor",
            u"Ungdjur",
            u"Lekpar",
            u"Mycket fina vildfångade"
        ]

        self.comments_plants = [
            u"Snabbväxande",
            u"Ljuskrävande",
            u"En rejäl bunt"
        ]

    def get_seller_name_and_email(self):
        first_name = random.choice(self.first_names)
        last_name = random.choice(self.surnames)
        name = "{} {}".format(first_name, last_name)
        email = "{}.{}@mail.com".format(first_name, last_name).lower()
        return name, email
    
    def get_address(self):
        city, zip_base = random.choice(self.cities)
        street = random.choice(self.street_names)
        zip_code = u"{}{} {}{}".format(zip_base, random.randint(0, 9), random.randint(0, 9), random.randint(0, 9))
        street_number = random.randint(1, 100)
        address = u"{} {}, {} {}".format(street, street_number, zip_code, city)
        return (address)

    def get_phone(self):
        phone_number = "0"
        for n in range(9):
            if n == 3:
                phone_number += "-"
            phone_number += str(random.randint(0, 9))
        return phone_number

    def get_password(self):
        return random.choice(self.passwords)
        # return u"password"

    def get_aquarium_club(self):
        return random.choice(self.aquarium_clubs)
    
    def get_animal_comment(self):
        comment = ""
        if random.randint(0, 100) < 25:
            comment =random.choice(self.comments_animals)
        return comment

    def get_plant_comment(self):
        comment = ""
        if random.randint(0, 100) < 25:
            comment = random.choice(self.comments_plants)
            return comment

    def generate_seller(self, admin="no"):
        name, email = self.get_seller_name_and_email()
        address = self.get_address()
        phone_number = self.get_phone()
        password = self.get_password()
        club = self.get_aquarium_club()
        return name, address, email, phone_number, club, admin, password, time.strftime("%Y-%m-%d %H:%M:%S"), "yes", "yes", "no", "no"

    def get_fish_name(self):
        return random.choice(self.fish_names)

    def get_plant_name(self):
        return random.choice(self.plant_names)

    def get_shrimp_name(self):
         return random.choice(self.shrimp_names)

    def generate_fish_auction_post(self, seller_id):
        type = 1
        sci_name, pop_name = self.get_fish_name()
        quantity = random.randint(1, 10) 
        comment = self.get_animal_comment()
        mimimum_price = None
        fixed_price = None
        if random.randint(0, 100) < 20:
            mimimum_price = random.randint(1, 10) * 10
        return seller_id, sci_name, pop_name, quantity, comment, type, time.strftime("%Y-%m-%d %H:%M:%S"), mimimum_price, fixed_price, "no", "no"

    def generate_plant_fleamarket_post(self, seller_id):
        type = 8
        comment = self.get_plant_comment()
        pop_name, sci_name = self.get_plant_name()
        quantity = random.randint(1, 10)
        mimimum_price = None
        fixed_price = random.randint(1, 10) * 10
        return seller_id, sci_name, pop_name, quantity, comment, type, time.strftime("%Y-%m-%d %H:%M:%S"), mimimum_price, fixed_price, "no", "no"

    def generate_shrimp_fleamarket_post(self, seller_id):
        type = 4
        comment = self.get_animal_comment()
        pop_name, sci_name = self.get_shrimp_name()
        quantity = random.randint(1, 10)
        mimimum_price = None
        fixed_price = random.randint(1, 10) * 10
        return seller_id, sci_name, pop_name, quantity, comment, type, time.strftime("%Y-%m-%d %H:%M:%S"), mimimum_price, fixed_price, "no", "no"

if __name__ == "__main__":
    
    F = Faker()
    # print(F.get_fish_name())
    print(F.get_seller_name_and_email())
    print(F.get_address())
    print(F.get_phone())
    print(F.generate_fish_auction_post(3))

